### PR TITLE
[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119107
* __->__ #118438

Part of implementing CacheEntry invalidation to fix https://github.com/pytorch/pytorch/issues/112090.

Changes:
- Move CacheEntry and ExtraState to C++
- Use pybind to control reference counting
- Use std::list instead of manually implementing a linked list

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/csrc/dynamo/cache_entry.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+static void cache_entry_dealloc(CacheEntry* e) {
+  Py_XDECREF(e->check_fn);
+  Py_XDECREF(e->code);
+  // This will recursively call cache_entry_dealloc for the next items in the
+  // linked list.
+  Py_XDECREF(e->next);
+  Py_TYPE(e)->tp_free((PyObject*)e);
+}
+
+PyTypeObject CacheEntryType = {
```
Comment (jansel) [inline torch/csrc/dynamo/cache_entry.cpp:101]: Consider rewriting this to use pybind.

File: torch/csrc/dynamo/cache_entry.cpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+CacheEntry* create_cache_entry(
+    CacheEntry* next,
+    PyObject* guarded_code) {
+  PyObject* check_fn = PyObject_GetAttrString(guarded_code, "check_fn"); // new reference
+  PyCodeObject* code = (PyCodeObject*)PyObject_GetAttrString(guarded_code, "code"); // new reference
+
+  // equivalent to CacheEntry(check_fn, code, next) in Python
+  PyObject* args = Py_BuildValue("OOO", check_fn, code, next);
+  CacheEntry* e = (CacheEntry*)PyObject_CallObject((PyObject*)&CacheEntryType, args); // new reference
```
Comment (jansel) [inline torch/csrc/dynamo/cache_entry.cpp:82]: Missing error checks on all these calls.

File: torch/csrc/dynamo/cache_entry.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+  static char *kwlist[] = {"check_fn", "code", "next", NULL};
+
+  int ret = PyArg_ParseTupleAndKeywords(
+    args, kwds, "OOO", kwlist,
+    &check_fn, &code, &next);
+
+  if (!ret) return -1;
+
+  if (check_fn) {
```
Comment (jansel) [inline torch/csrc/dynamo/cache_entry.cpp:51]: When will this be false?  Shouldn't that be an error?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/csrc/dynamo/cache_entry.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+*/
+
+typedef struct CacheEntry CacheEntry;
+typedef struct ExtraState ExtraState;
+
+#ifdef __cplusplus
+
+typedef struct CacheEntry {
+  // check the guards: lambda: <locals of user function>: bool
+  py::object check_fn;
```
Comment (jansel) [inline torch/csrc/dynamo/cache_entry.h:42]: Have you benchmarked this?  This (and code) are the only part on the critical path, so want to much sure making it a py::object doesn't slow things down.

You can use ./benchmarks/dynamo/microbenchmarks/overheads.py

**Review by jansel — CHANGES_REQUESTED** [review]
Test failures?

Other than minor comment and test failures looks good.

File: torch/csrc/dynamo/cache_entry.h (hunk 4 above)
Comment (williamwen42) [reply torch/csrc/dynamo/cache_entry.h:42]: My results are:
```
pre changes
eager    6.4us (warmup=0.0s)
compiled 15.0us (warmup=8.3s)

post changes
eager    6.2us (warmup=0.0s)
compiled 14.9us (warmup=9.1s)
```

Comment (jansel) [reply torch/csrc/dynamo/cache_entry.h:42]: Yes seems fine, so just need to address test failures.

File: .lintrunner.toml [hunk 5]
```diff
@@ -463,6 +463,7 @@ include_patterns = [
 ]
 exclude_patterns = [
     'torch/csrc/utils/pybind.h',
+    'torch/csrc/dynamo/**/*',
```
Comment (jansel) [inline .lintrunner.toml]: Why?  Can we just fix the lints?

**Review by jansel — CHANGES_REQUESTED** [review]

File: .lintrunner.toml (hunk 5 above)
Comment (williamwen42) [reply .lintrunner.toml]: I was experimenting to see if including `torch/csrc/utils/pybind.h` over `pybind/pybind11.h` was causing issues in CI. I don't think this is the case - I've reverted the lintrunner settings.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**